File: packages/libs/SubstrateML/Sources/SubstrateML/ConceptImplications.swift
// ConceptImplications.swift
//
// Sound logical implications over a `FormalContext` via bounded
// Duquenne–Guigues canonical basis enumeration (Next-Closure).
//
// ALGORITHM: Next-Closure (Ganter 1999, "Two Basic Algorithms in Concept
// Analysis", TH Darmstadt Preprint 831). Enumeration visits candidate
// attribute sets in lectic order over the sorted attribute universe and
// identifies pseudo-intents: sets A whose closure A'' strictly contains A.
// Every pseudo-intent A yields one implication A → (A'' \ A) in the
// canonical basis. The basis is minimal (Duquenne–Guigues): no proper
// subset of it implies the same consequences, and every sound implication
// over the context follows from the basis by Armstrong's rules.
//
// SOUNDNESS CONTRACT:
// Every emitted implication (premise → conclusion) holds universally across
// every row of the context: if a row carries every attribute in premise, it
// carries every attribute in conclusion. This is guaranteed because
// conclusion = closure(premise) \ premise — the closure operator returns
// exactly the attributes shared by every row carrying premise.
//
// INCOMPLETENESS CONTRACT (bounding):
// Pseudo-intent enumeration is NP-hard in the worst case (exponential in
// the number of attributes). Two caps terminate the engine early:
//   - `maxImplications`: emits at most this many implications. When the cap
//     hits, `isTruncated == true` and the returned basis may be incomplete
//     (some pseudo-intents were not reached).
//   - `maxPremiseSize`: pseudo-intents with |premise| > this cap are
//     silently skipped (not emitted, not counted toward maxImplications).
//     This lets callers exclude complex implications without halting
//     enumeration entirely. No soundness violation: only emission, not
//     correctness, is affected.
//
// Together these make the algorithm tractable for estate-scale contexts.
// The `isTruncated` flag tells callers whether the basis is the complete
// canonical basis (false) or a sound prefix of it (true).
//
// DETERMINISM: the attribute universe is sorted (FormalAttribute is
// Comparable by lexicographic (namespace, key, value) order), so the
// lectic order and therefore the enumeration order are fully determined
// by the context. Same input → same implications in the same order.
//
// The Rust mirror lives at:
//   packages/libs/SubstrateML/rust/src/concept_implications.rs
// Both versions must produce bit-identical output on the shared conformance
// vectors in:
//   docs/engineering/substrate_reference/test-harness/vectors/concept_implications.json

import Foundation

// MARK: - Implication

/// One sound logical implication: if a row carries every attribute in
/// `premise`, it carries every attribute in `conclusion` as well.
///
/// The `premise` and `conclusion` are disjoint by construction:
/// `conclusion == closure(premise) \ premise` — the closure-minus-premise
/// delta. The union `premise ∪ conclusion` is a closed set (a formal
/// concept intent or union of concept intents) in the context.
///
/// This is a genuine logical implication — not a structural cover delta.
/// Every row in the context that satisfies the premise must satisfy the
/// conclusion, without exception.
public struct Implication: Hashable, Codable, Sendable {
    /// The premise attribute set. Non-empty for all canonical-basis
    /// implications (the empty premise would force every row to carry every
    /// attribute, which only holds when all attributes are global).
    public let premise: Set<FormalAttribute>
    /// The conclusion: `closure(premise) \ premise`. Always non-empty
    /// (sets with closure equal to themselves are concepts, not pseudo-intents,
    /// and are not emitted as implications).
    public let conclusion: Set<FormalAttribute>

    /// Creates an implication with the given premise and conclusion.
    ///
    /// The caller is responsible for the invariant `premise ∩ conclusion == ∅`.
    /// The `ConceptImplications` engine enforces this by construction.
    public init(premise: Set<FormalAttribute>, conclusion: Set<FormalAttribute>) {
        self.premise = premise
        self.conclusion = conclusion
    }
}

// MARK: - ConceptImplications

/// Sound logical implications over a `FormalContext` via bounded
/// Duquenne–Guigues canonical basis enumeration.
///
/// The basis is *sound*: every emitted implication holds universally in
/// the context. The basis may be *incomplete* when bounding caps bind
/// (`isTruncated == true`): some pseudo-intents were not reached, so the
/// returned set is a sound prefix of the full canonical basis.
///
/// Compute via `ConceptImplications.conceptImplications(over:context:maxImplications:maxPremiseSize:)`.
///
/// See the file-header comment for the algorithm, soundness/incompleteness
/// contracts, and the cite to Ganter (1999).
public struct ConceptImplications: Sendable {
    /// The emitted implications, sorted for determinism.
    ///
    /// Sort order: premise size ascending, then lexicographic premise
    /// (sorted element list), then lexicographic conclusion. Fully specified
    /// so equal inputs yield identical output across runs and across the
    /// Swift and Rust implementations.
    public let implications: [Implication]

    /// True when a bounding cap (`maxImplications` or similar) terminated
    /// enumeration before all pseudo-intents were visited. When true, the
    /// basis is sound but may be incomplete.
    public let isTruncated: Bool

    /// Creates a `ConceptImplications` result.
    public init(implications: [Implication], isTruncated: Bool) {
        self.implications = implications
        self.isTruncated = isTruncated
    }

    // MARK: - Engine

    /// Compute the bounded Duquenne–Guigues canonical basis for `context`.
    ///
    /// Uses Next-Closure (Ganter 1999) to enumerate pseudo-intents in
    /// lectic order over the context's sorted attribute universe. For each
    /// pseudo-intent A (where `A'' ≠ A`), emits
    /// `Implication(premise: A, conclusion: A'' \ A)`.
    ///
    /// Bounding:
    /// - `maxImplications`: hard cap on the number of emitted implications.
    ///   When reached, sets `isTruncated = true` and terminates early.
    /// - `maxPremiseSize`: pseudo-intents with `|A| > maxPremiseSize` are
    ///   skipped (not emitted, enumeration continues). This is a filter,
    ///   not a terminator — `isTruncated` remains `false` unless the
    ///   `maxImplications` cap is hit.
    ///
    /// Empty context returns empty implications without truncation.
    ///
    /// - Parameters:
    ///   - concepts: the concept set from `BoundedConceptMiner`. Passed
    ///     for API symmetry with `ConceptCoverDeltas`; only `context` is
    ///     used by the engine. Pass `[]` when a full-context computation
    ///     is desired without concept-set filtering.
    ///   - context: the fully-materialised `FormalContext`.
    ///   - maxImplications: maximum number of implications to emit.
    ///     Values ≤ 0 are treated as 0 (empty result, not truncated unless
    ///     pseudo-intents exist — use `Int.max` for an uncapped run).
    ///   - maxPremiseSize: maximum premise size to emit. Pseudo-intents
    ///     larger than this are skipped silently.
    /// - Returns: `ConceptImplications` with the emitted basis and
    ///   truncation flag.
    public static func conceptImplications(
        over concepts: [FormalConcept],
        context: FormalContext,
        maxImplications: Int,
        maxPremiseSize: Int
    ) -> ConceptImplications {
        // Empty context → empty basis, not truncated.
        guard context.attributes.count > 0, context.rowCount > 0 else {
            return ConceptImplications(implications: [], isTruncated: false)
        }

        // Zero cap means: return nothing (but do not signal truncation when
        // there are no pseudo-intents to emit). We check this by running the
        // engine with a zero cap — if it would have emitted something it will
        // set isTruncated.
        //
        // For maxImplications <= 0 with attributes present: run the normal
        // path so we correctly set isTruncated.
        return nextClosureEngine(
            context: context,
            maxImplications: maxImplications,
            maxPremiseSize: maxPremiseSize
        )
    }

    // MARK: - D-G Basis Engine (size-ordered pseudo-intent enumeration)

    /// Computes the Duquenne–Guigues canonical basis by enumerating all
    /// subsets of the attribute universe in size order (0, 1, 2, ...,
    /// maxPremiseSize) and testing each for pseudo-intenthood against the
    /// implication set L built incrementally.
    ///
    /// CORRECTNESS: this algorithm correctly computes the minimal D-G basis.
    /// A set P is a D-G pseudo-intent iff:
    ///   (1) closure_context(P) ≠ P  (P is not a concept intent / closed set)
    ///   (2) For every D-G pseudo-intent Q with Q ⊊ P: Q'' ⊆ P
    ///       (all conclusions of smaller pseudo-intents are already in P)
    ///
    /// By processing in non-decreasing size order, when we test P, all
    /// pseudo-intents of size < |P| are already in L. Condition (2) is
    /// then equivalent to: for all (premise, conclusion) in L where
    /// premise ⊊ P: conclusion ⊆ P.
    ///
    /// BOUNDING: subsets of size > maxPremiseSize are not tested. We still
    /// record each found pseudo-intent in stemBase (even when maxPremiseSize
    /// causes it to be skipped from emission) so that minimality checks for
    /// SUBSEQUENT larger subsets remain correct. Emission is bounded
    /// separately by maxImplications.
    ///
    /// COMPLEXITY: O(C(n, maxPremiseSize) × k) where n = |universe| and
    /// k = number of pseudo-intents found (for each candidate, checking
    /// condition 2 is O(k × |P|)). Tractable for typical estate contexts
    /// with maxPremiseSize ≤ 10.
    private static func nextClosureEngine(
        context: FormalContext,
        maxImplications: Int,
        maxPremiseSize: Int
    ) -> ConceptImplications {
        let universe = context.attributes // sorted ascending
        let n = universe.count

        var emitted: [Implication] = []
        // stemBase: all found D-G pseudo-intents and their conclusions,
        // including those with |premise| > maxPremiseSize (not emitted but
        // still needed for minimality checks of larger candidates).
        var stemBase: [(premise: Set<FormalAttribute>, conclusion: Set<FormalAttribute>)] = []
        var isTruncated = false

        // Enumerate by size from 0 to min(n, maxPremiseSize).
        // Use an index-set representation to enumerate combinations.
        let capSize = min(n, maxPremiseSize)

        // Special case: size 0 (empty set).
        let emptyContextClosure = Set(context.closure(of: []))
        if !emptyContextClosure.isEmpty {
            // {} is a pseudo-intent: closure of {} is non-empty (global attrs).
            // Condition (2): no proper subset of {} is a pseudo-intent. ✓
            if maxImplications <= 0 {
                isTruncated = true
                return ConceptImplications(implications: [], isTruncated: true)
            }
            let premise = Set<FormalAttribute>()
            let conclusion = emptyContextClosure
            emitted.append(Implication(premise: premise, conclusion: conclusion))
            stemBase.append((premise: premise, conclusion: conclusion))
            if emitted.count >= maxImplications {
                isTruncated = true
                return ConceptImplications(
                    implications: sortedImplications(emitted),
                    isTruncated: true
                )
            }
        }

        // Enumerate subsets of size 1..capSize via combination index arrays.
        // indices[k] ∈ [0, n) with indices[0] < indices[1] < ... < indices[k-1].
        // Skip if capSize is 0 (maxPremiseSize == 0: only the empty set is checked).
        guard capSize >= 1 else {
            return ConceptImplications(implications: sortedImplications(emitted), isTruncated: false)
        }
        for size in 1...capSize {
            guard size <= n else { break }
            // Enumerate all C(n, size) combinations of universe indices.
            var indices = Array(0..<size) // starting combination
            while true {
                // Build the subset from indices.
                let premiseSet = Set(indices.map { universe[$0] })

                // Test pseudo-intenthood.
                if isDGPseudoIntent(
                    premiseSet: premiseSet,
                    context: context,
                    stemBase: stemBase
                ) {
                    let contextClosureSet = Set(context.closure(of: Array(premiseSet)))
                    let conclusion = contextClosureSet.subtracting(premiseSet)
                    // Always record in stemBase for minimality (even if not emitted).
                    stemBase.append((premise: premiseSet, conclusion: conclusion))

                    // Emit if size ≤ maxPremiseSize (it always is here, since
                    // we cap the outer loop at capSize = min(n, maxPremiseSize)).
                    if maxImplications <= 0 {
                        isTruncated = true
                        return ConceptImplications(
                            implications: sortedImplications(emitted),
                            isTruncated: true
                        )
                    }
                    emitted.append(Implication(premise: premiseSet, conclusion: conclusion))
                    if emitted.count >= maxImplications {
                        // Check if more pseudo-intents exist beyond this point.
                        isTruncated = hasMorePseudoIntents(
                            after: indices, size: size,
                            universe: universe, context: context, stemBase: stemBase,
                            capSize: capSize
                        )
                        return ConceptImplications(
                            implications: sortedImplications(emitted),
                            isTruncated: isTruncated
                        )
                    }
                }

                // Advance to next combination in colexicographic order.
                guard let next = nextCombination(indices: indices, n: n) else { break }
                indices = next
            }
        }

        return ConceptImplications(
            implications: sortedImplications(emitted),
            isTruncated: false
        )
    }

    /// Tests whether `premiseSet` is a D-G pseudo-intent w.r.t. the context
    /// and the current stem base.
    ///
    /// Conditions:
    ///   (1) closure_context(premiseSet) ≠ premiseSet  (not closed)
    ///   (2) For every (premise, conclusion) in stemBase where premise ⊊ premiseSet:
    ///       conclusion ⊆ premiseSet  (every smaller pseudo-intent's conclusion
    ///       is already contained in premiseSet)
    private static func isDGPseudoIntent(
        premiseSet: Set<FormalAttribute>,
        context: FormalContext,
        stemBase: [(premise: Set<FormalAttribute>, conclusion: Set<FormalAttribute>)]
    ) -> Bool {
        // Condition (1): not closed.
        let contextClosure = Set(context.closure(of: Array(premiseSet)))
        guard contextClosure != premiseSet else { return false }

        // Condition (2): all smaller pseudo-intent conclusions are in premiseSet.
        for (premise, conclusion) in stemBase {
            // Only check strict subsets.
            if premise.count < premiseSet.count, premise.isSubset(of: premiseSet) {
                if !conclusion.isSubset(of: premiseSet) {
                    return false
                }
            }
        }
        return true
    }

    /// Returns the next combination of `size` indices from [0, n) in
    /// lexicographic order after `indices`, or `nil` when `indices` is the
    /// last combination.
    private static func nextCombination(indices: [Int], n: Int) -> [Int]? {
        var result = indices
        let k = result.count
        // Find the rightmost index that can be incremented.
        var i = k - 1
        while i >= 0 {
            if result[i] < n - (k - i) {
                result[i] += 1
                // Reset all indices to the right.
                for j in (i + 1)..<k {
                    result[j] = result[j - 1] + 1
                }
                return result
            }
            i -= 1
        }
        return nil // all combinations exhausted
    }

    /// Checks if there are any more pseudo-intents after the current state.
    ///
    /// Used to set `isTruncated` correctly when the maxImplications cap hits.
    /// Returns `true` if at least one more pseudo-intent exists in the
    /// remaining enumeration range.
    private static func hasMorePseudoIntents(
        after currentIndices: [Int],
        size: Int,
        universe: [FormalAttribute],
        context: FormalContext,
        stemBase: [(premise: Set<FormalAttribute>, conclusion: Set<FormalAttribute>)],
        capSize: Int
    ) -> Bool {
        let n = universe.count

        // Continue within the current size.
        var indices = currentIndices
        while let next = nextCombination(indices: indices, n: n) {
            indices = next
            let premiseSet = Set(indices.map { universe[$0] })
            if isDGPseudoIntent(premiseSet: premiseSet, context: context, stemBase: stemBase) {
                return true
            }
        }

        // Check larger sizes (only up to capSize).
        guard size + 1 <= capSize else { return false }
        for nextSize in (size + 1)...capSize {
            guard nextSize <= n else { break }
            var idx = Array(0..<nextSize)
            while true {
                let premiseSet = Set(idx.map { universe[$0] })
                if isDGPseudoIntent(premiseSet: premiseSet, context: context, stemBase: stemBase) {
                    return true
                }
                guard let next = nextCombination(indices: idx, n: n) else { break }
                idx = next
            }
        }
        return false
    }

    /// Sorts emitted implications for determinism.
    ///
    /// Order: premise size ascending, lexicographic premise elements,
    /// lexicographic conclusion elements. This order is fully specified
    /// and identical across Swift and Rust.
    private static func sortedImplications(_ implications: [Implication]) -> [Implication] {
        return implications.sorted { lhs, rhs in
            if lhs.premise.count != rhs.premise.count {
                return lhs.premise.count < rhs.premise.count
            }
            let lp = lhs.premise.sorted(), rp = rhs.premise.sorted()
            if lp != rp { return lp.lexicographicallyPrecedes(rp) }
            return lhs.conclusion.sorted()
                .lexicographicallyPrecedes(rhs.conclusion.sorted())
        }
    }

}

File: packages/libs/SubstrateML/rust/src/concept_implications.rs
//! Sound logical implications over a `FormalContext` via bounded
//! Duquenne–Guigues canonical basis enumeration.
//!
//! # Algorithm
//!
//! Size-ordered pseudo-intent enumeration (Ganter & Wille, "Formal Concept
//! Analysis: Mathematical Foundations", Chapter 7). The Rust implementation
//! is a direct port of `ConceptImplications.swift` and must produce
//! bit-identical output on the shared conformance vectors in
//! `docs/engineering/substrate_reference/test-harness/vectors/concept_implications.json`.
//!
//! A set P is a Duquenne–Guigues pseudo-intent iff:
//!   (1) closure_context(P) ≠ P  (P is not a concept intent / closed set)
//!   (2) For every D-G pseudo-intent Q with Q ⊊ P: Q'' ⊆ P
//!       (all conclusions from smaller pseudo-intents fit inside P)
//!
//! The algorithm enumerates all subsets of size 0..maxPremiseSize in
//! non-decreasing size order (lexicographic within each size). For each
//! subset P, conditions (1) and (2) are tested against the implication set
//! built so far. When P passes both conditions it is a pseudo-intent and
//! yields `Implication { premise: P, conclusion: closure(P) \ P }`.
//!
//! # Soundness
//!
//! Every emitted implication holds in the context: if a row carries every
//! attribute in `premise`, it carries every attribute in `conclusion`.
//! This follows because `conclusion = closure_context(premise) \ premise`,
//! and the closure operator returns exactly the attributes shared by every
//! row carrying `premise`.
//!
//! # Bounding
//!
//! Two caps limit the output:
//! - `max_implications`: hard cap on the number of emitted implications.
//!   When reached, `is_truncated = true`.
//! - `max_premise_size`: pseudo-intents with `|premise| > max_premise_size`
//!   are silently skipped (not emitted, enumeration continues). This is a
//!   filter, not a terminator; it does not set `is_truncated` by itself.
//!
//! # Rust–Swift contract
//!
//! The attribute universe is sorted by the `Ord` on `FormalAttribute`
//! (lexicographic over `(namespace, key, value)`), identical to the Swift
//! `Comparable` order. The combination enumeration visits subsets in
//! lexicographic order on the sorted universe — identical to Swift.
//! Implication output is sorted: premise size ascending, lex premise,
//! lex conclusion.

use std::collections::BTreeSet;
use crate::formal_concept_analysis::{FormalAttribute, FormalContext};

// MARK: - Types

/// One sound logical implication: every row carrying all attributes in
/// `premise` also carries all attributes in `conclusion`.
///
/// `premise` and `conclusion` are disjoint: `conclusion` is
/// `closure_context(premise) \ premise`. Both are stored as sorted
/// `Vec<FormalAttribute>` for deterministic output.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Implication {
    /// The premise attribute set (sorted ascending).
    pub premise: Vec<FormalAttribute>,
    /// The conclusion: `closure(premise) \ premise` (sorted ascending).
    pub conclusion: Vec<FormalAttribute>,
}

/// Sound logical implications over a `FormalContext` via bounded
/// Duquenne–Guigues canonical basis enumeration.
///
/// The basis is *sound*: every emitted implication holds universally.
/// The basis may be *incomplete* when bounding caps bind
/// (`is_truncated == true`).
#[derive(Debug, Clone, PartialEq)]
pub struct ConceptImplications {
    /// Emitted implications, sorted for determinism.
    pub implications: Vec<Implication>,
    /// True when the `max_implications` cap terminated enumeration early.
    pub is_truncated: bool,
}

impl ConceptImplications {
    /// Computes the bounded D-G canonical basis for `context`.
    ///
    /// Uses size-ordered pseudo-intent enumeration. Subsets of the attribute
    /// universe are tested in non-decreasing size order (lexicographic within
    /// each size) against the D-G pseudo-intent conditions. Sound implications
    /// are emitted and accumulated into the stem base for minimality checking
    /// of subsequent candidates.
    ///
    /// # Parameters
    ///
    /// - `context`: the fully-materialised `FormalContext`.
    /// - `max_implications`: maximum number of implications to emit. Set to
    ///   `usize::MAX` for uncapped enumeration.
    /// - `max_premise_size`: maximum premise size to emit. Pseudo-intents
    ///   with `|premise| > max_premise_size` are skipped silently.
    ///
    /// Returns an empty basis when the context has no rows or no attributes.
    pub fn compute(
        context: &FormalContext,
        max_implications: usize,
        max_premise_size: usize,
    ) -> ConceptImplications {
        let universe = context.attributes();
        let n = universe.len();

        if n == 0 || context.row_count() == 0 {
            return ConceptImplications { implications: Vec::new(), is_truncated: false };
        }

        // stem_base accumulates all found pseudo-intents (premise, conclusion)
        // including those not emitted due to max_premise_size.
        let mut stem_base: Vec<(BTreeSet<FormalAttribute>, BTreeSet<FormalAttribute>)> = Vec::new();
        let mut emitted: Vec<Implication> = Vec::new();
        let mut is_truncated = false;

        let cap_size = n.min(max_premise_size);

        // Size 0: test the empty set.
        {
            let premise = BTreeSet::new();
            let closure = context_closure_set(context, &premise);
            if !closure.is_empty() {
                // {} is a pseudo-intent (global attributes exist).
                let conclusion: BTreeSet<FormalAttribute> = closure.difference(&premise).cloned().collect();
                stem_base.push((premise.clone(), conclusion.clone()));

                if max_implications == 0 {
                    return ConceptImplications {
                        implications: Vec::new(),
                        is_truncated: true,
                    };
                }
                emitted.push(Implication {
                    premise: sorted_vec(&premise),
                    conclusion: sorted_vec(&conclusion),
                });
                if emitted.len() >= max_implications {
                    let has_next = has_more_pseudo_intents_after(
                        &[], 0, context, &stem_base, cap_size,
                    );
                    return ConceptImplications {
                        implications: sort_implications(emitted),
                        is_truncated: has_next,
                    };
                }
            }
        }

        // Sizes 1..cap_size.
        if cap_size == 0 {
            return ConceptImplications {
                implications: sort_implications(emitted),
                is_truncated: false,
            };
        }

        for size in 1..=cap_size {
            // Enumerate all C(n, size) combinations of attribute indices.
            let mut indices: Vec<usize> = (0..size).collect();
            loop {
                // Build premise set from current combination.
                let premise: BTreeSet<FormalAttribute> = indices.iter()
                    .map(|&i| universe[i].clone())
                    .collect();

                if is_dg_pseudo_intent(&premise, context, &stem_base) {
                    let closure = context_closure_set(context, &premise);
                    let conclusion: BTreeSet<FormalAttribute> = closure.difference(&premise).cloned().collect();
                    stem_base.push((premise.clone(), conclusion.clone()));

                    if max_implications == 0 {
                        is_truncated = true;
                        break;
                    }
                    emitted.push(Implication {
                        premise: sorted_vec(&premise),
                        conclusion: sorted_vec(&conclusion),
                    });
                    if emitted.len() >= max_implications {
                        let has_next = has_more_pseudo_intents_after(
                            &indices, size, context, &stem_base, cap_size,
                        );
                        return ConceptImplications {
                            implications: sort_implications(emitted),
                            is_truncated: has_next,
                        };
                    }
                }

                // Advance to next combination in lexicographic order.
                match next_combination(&mut indices, n) {
                    true => {}
                    false => break,
                }
            }
            if is_truncated { break; }
        }

        ConceptImplications {
            implications: sort_implications(emitted),
            is_truncated,
        }
    }
}

// MARK: - Internal helpers

/// Tests whether `premise` is a D-G pseudo-intent w.r.t. the context
/// and the current stem base.
///
/// Conditions:
///   (1) closure_context(premise) ≠ premise  (not closed)
///   (2) For every (p, q) in stem_base where p ⊊ premise: q ⊆ premise
fn is_dg_pseudo_intent(
    premise: &BTreeSet<FormalAttribute>,
    context: &FormalContext,
    stem_base: &[(BTreeSet<FormalAttribute>, BTreeSet<FormalAttribute>)],
) -> bool {
    // Condition (1): not closed.
    let closure = context_closure_set(context, premise);
    if closure == *premise {
        return false;
    }

    // Condition (2): all conclusions of strict-subset pseudo-intents are in premise.
    for (p, q) in stem_base {
        if p.len() < premise.len() && p.is_subset(premise) {
            if !q.is_subset(premise) {
                return false;
            }
        }
    }
    true
}

/// Computes the context closure of `attrs` as a `BTreeSet`.
///
/// Uses `FormalContext::closure` (which returns a sorted `Vec`) and converts
/// to a `BTreeSet` for set operations. Mirrors the Swift
/// `context.closure(of:)` call.
fn context_closure_set(
    context: &FormalContext,
    attrs: &BTreeSet<FormalAttribute>,
) -> BTreeSet<FormalAttribute> {
    let attrs_vec: Vec<FormalAttribute> = attrs.iter().cloned().collect();
    context.closure(&attrs_vec).into_iter().collect()
}

/// Advances `indices` to the next combination of `size` values from [0, n)
/// in lexicographic order. Returns `true` on success, `false` when `indices`
/// is the last combination.
fn next_combination(indices: &mut Vec<usize>, n: usize) -> bool {
    let k = indices.len();
    let mut i = k;
    loop {
        if i == 0 {
            return false;
        }
        i -= 1;
        if indices[i] < n - (k - i) {
            indices[i] += 1;
            for j in (i + 1)..k {
                indices[j] = indices[j - 1] + 1;
            }
            return true;
        }
    }
}

/// Checks whether any more pseudo-intents exist after the current position.
/// Used to set `is_truncated` correctly when the `max_implications` cap hits.
fn has_more_pseudo_intents_after(
    current_indices: &[usize],
    current_size: usize,
    context: &FormalContext,
    stem_base: &[(BTreeSet<FormalAttribute>, BTreeSet<FormalAttribute>)],
    cap_size: usize,
) -> bool {
    let n = context.attributes().len();
    let universe = context.attributes();

    // Continue within the current size.
    if current_size > 0 {
        let mut indices = current_indices.to_vec();
        while next_combination(&mut indices, n) {
            let premise: BTreeSet<FormalAttribute> = indices.iter()
                .map(|&i| universe[i].clone())
                .collect();
            if is_dg_pseudo_intent(&premise, context, stem_base) {
                return true;
            }
        }
    }

    // Check larger sizes.
    for next_size in (current_size + 1)..=cap_size {
        if next_size > n { break; }
        let mut indices: Vec<usize> = (0..next_size).collect();
        loop {
            let premise: BTreeSet<FormalAttribute> = indices.iter()
                .map(|&i| universe[i].clone())
                .collect();
            if is_dg_pseudo_intent(&premise, context, stem_base) {
                return true;
            }
            if !next_combination(&mut indices, n) { break; }
        }
    }
    false
}

/// Converts a `BTreeSet<FormalAttribute>` to a sorted `Vec<FormalAttribute>`.
/// Since `BTreeSet` is already sorted, this is a direct collect.
fn sorted_vec(set: &BTreeSet<FormalAttribute>) -> Vec<FormalAttribute> {
    set.iter().cloned().collect()
}

/// Sorts a vector of `Implication` for deterministic output.
///
/// Order: premise size ascending, lex premise, lex conclusion.
/// Identical to the Swift `sortedImplications` function.
fn sort_implications(mut implications: Vec<Implication>) -> Vec<Implication> {
    implications.sort_by(|a, b| {
        a.premise.len().cmp(&b.premise.len())
            .then_with(|| a.premise.cmp(&b.premise))
            .then_with(|| a.conclusion.cmp(&b.conclusion))
    });
    implications
}

// MARK: - Tests

#[cfg(test)]
mod tests {
    use super::*;
    use crate::formal_concept_analysis::FormalContext;
    // HashSet is test-only (premise/conclusion set comparisons below);
    // importing it at module scope would warn on non-test builds.
    use std::collections::HashSet;

    fn a() -> FormalAttribute { FormalAttribute::new("t", "k", "a") }
    fn b() -> FormalAttribute { FormalAttribute::new("t", "k", "b") }
    fn c() -> FormalAttribute { FormalAttribute::new("t", "k", "c") }

    // ci-001: empty context.
    #[test]
    fn ci001_empty_context() {
        let ctx = FormalContext::new(&[]);
        let result = ConceptImplications::compute(&ctx, 100, 100);
        assert!(result.implications.is_empty());
        assert!(!result.is_truncated);
    }

    // ci-002: two-row context. Row0:{a}, Row1:{b,c}. D-G basis: {b}->{c}, {c}->{b}.
    #[test]
    fn ci002_two_row_three_attribute() {
        let ctx = FormalContext::new(&[
            vec![a()],
            vec![b(), c()],
        ]);
        let result = ConceptImplications::compute(&ctx, 100, 100);
        assert!(!result.is_truncated);
        assert_eq!(result.implications.len(), 2);
        // {b} -> {c}
        assert_eq!(result.implications[0].premise, vec![b()]);
        assert_eq!(result.implications[0].conclusion, vec![c()]);
        // {c} -> {b}
        assert_eq!(result.implications[1].premise, vec![c()]);
        assert_eq!(result.implications[1].conclusion, vec![b()]);
    }

    // ci-003: three singletons, maxImplications=2. Truncated.
    #[test]
    fn ci003_cap_truncation() {
        let ctx = FormalContext::new(&[
            vec![a()],
            vec![b()],
            vec![c()],
        ]);
        let result = ConceptImplications::compute(&ctx, 2, 100);
        assert!(result.is_truncated);
        assert_eq!(result.implications.len(), 2);
        // First two implications in size-lex order: {a,b}->{c}, {a,c}->{b}.
        assert_eq!(result.implications[0].premise, vec![a(), b()]);
        assert_eq!(result.implications[0].conclusion, vec![c()]);
        assert_eq!(result.implications[1].premise, vec![a(), c()]);
        assert_eq!(result.implications[1].conclusion, vec![b()]);
    }

    // ci-004: global attribute. Row0:{a,b}, Row1:{b,c}. D-G basis: {}->{b}.
    #[test]
    fn ci004_global_attribute() {
        let ctx = FormalContext::new(&[
            vec![a(), b()],
            vec![b(), c()],
        ]);
        let result = ConceptImplications::compute(&ctx, 100, 100);
        assert!(!result.is_truncated);
        assert_eq!(result.implications.len(), 1);
        assert_eq!(result.implications[0].premise, Vec::<FormalAttribute>::new());
        assert_eq!(result.implications[0].conclusion, vec![b()]);
    }

    // Soundness: for every emitted implication (premise -> conclusion),
    // every row that carries all of premise also carries all of conclusion.
    #[test]
    fn soundness_holds_for_every_row() {
        let ctx = FormalContext::new(&[
            vec![a()],
            vec![b(), c()],
        ]);
        let result = ConceptImplications::compute(&ctx, 100, 100);
        let rows: Vec<std::collections::HashSet<FormalAttribute>> = vec![
            vec![a()].into_iter().collect(),
            vec![b(), c()].into_iter().collect(),
        ];
        for impl_ in &result.implications {
            let premise_set: HashSet<FormalAttribute> = impl_.premise.iter().cloned().collect();
            let conclusion_set: HashSet<FormalAttribute> = impl_.conclusion.iter().cloned().collect();
            for row in &rows {
                if premise_set.is_subset(row) {
                    assert!(
                        conclusion_set.is_subset(row),
                        "Soundness violation: row {:?} satisfies premise {:?} but not conclusion {:?}",
                        row, impl_.premise, impl_.conclusion
                    );
                }
            }
        }
    }

    // Determinism: two runs produce identical output.
    #[test]
    fn determinism() {
        let ctx = FormalContext::new(&[
            vec![a()],
            vec![b(), c()],
        ]);
        let first = ConceptImplications::compute(&ctx, 100, 100);
        let second = ConceptImplications::compute(&ctx, 100, 100);
        assert_eq!(first, second);
    }

    // maxPremiseSize filter: three singletons context with maxPremiseSize=1.
    // All D-G implications have size-2 premises -> none emitted, not truncated.
    #[test]
    fn max_premise_size_filters() {
        let ctx = FormalContext::new(&[
            vec![a()],
            vec![b()],
            vec![c()],
        ]);
        let result = ConceptImplications::compute(&ctx, 100, 1);
        assert!(result.implications.is_empty());
        assert!(!result.is_truncated);
    }

    // maxImplications=0: truncated when pseudo-intents exist.
    #[test]
    fn max_implications_zero_truncates() {
        let ctx = FormalContext::new(&[
            vec![a()],
            vec![b(), c()],
        ]);
        let result = ConceptImplications::compute(&ctx, 0, 100);
        assert!(result.implications.is_empty());
        assert!(result.is_truncated);
    }
}
